### PR TITLE
Memorize GenerationID

### PR DESCRIPTION
**Pull request by moolitayer**
Currently if at the time of the request, the provided genID is different then the current genID, every pushInterval (5 sec default) all messages are replied. Updated description.

[1] that would happen when prometheus-alert-buffer goes down & up and a client attempts to reconnect based on last known generation & index.

### CODE REVIEW

**Comment by moolitayer** [comment]
@ilackarms please review

**Comment by ilackarms** [comment]
I believe this will solve the issue; however I realized that the alternative is to simply specify the genID in the  `?generationID=<id>` it should have the same effect

you can get the generationID from a single `GET` and then use that to inform your watch

**Comment by moolitayer** [comment]
Sorry @ilackarms I'm not sure what you mean can you clarify?

**Comment by ilackarms** [comment]
@moolitayer the reason you're getting repeated messages is because you're not specifying the `generationID` paramter in the `GET` request which initiates the watch. 

To prevent repeat messages.
1. Do a `GET /topics/{topic}`. Note the `generationID` in the response object.
2. Connect to `/topics/{topic}/watch?generationID=<generation id>` to start the watch

**Comment by ilackarms** [comment]
although I'm not sure it's really a *feature* that we repeat messages when generationID is not specified by the consumer. Maybe this change is fine anyway. Thoughts @juliusv?

**Comment by moolitayer** [comment]
Thanks for the clarification @ilackarms 
currently If i'm starting a watch with an old generation I will get all the history every 5 sec while if i'm starting with the current generation I will only get the new messages.

**Comment by ilackarms** [comment]
the generation id is only generated when connecting to a fresh DB. are you seeing multiple generation IDs?

**Comment by ilackarms** [comment]
from the readme:

>The `generationID` query parameter is expected to match the `generationID` value returned from any `/topics/*` requests. If it does not match, all entries are returned instead of just the ones starting from `fromIndex`. The generation ID is created when the tool's database is first initialized.

In the case of a watch I actually believe it makes sense to "remember' the generation ID. In short, I agree with making this change

**Comment by juliusv** [comment]
Yeah, initially this tool had no streaming capabilities (only normal HTTP GET), and the way a client could make sure that they would only get events that they hadn't seen yet was to include both and index and the generation ID with their requests, where the generation ID ensures that you are only referring to an index that is valid in the current database (if the database gets wiped and reinitialized, you would otherwise miss some events in there if you are sending a high index from a previous DB).

For streaming, I think it makes sense to apply the generation ID + index from the request for the initial batch of messages, but then while waiting, I don't think you should be repeating already sent messages on the same stream even if the generation ID is not provided.

So in short, I think this change is a good idea :)

**Comment by moolitayer** [comment]
@simon3z @zgalor review/merge ?

**Comment by simon3z** [comment]
@moolitayer IIUC when the `genID` (the one provided from the http request) is `!=` from the `msgsResponse.GenerationID` then we keep re-sending all the alerts? Is that the bug?

**Comment by moolitayer** [comment]
if at the time of the request, the provided `genID` is different then the current generation[1], every `pushInterval` (5 sec default) all messages are replied. Updated description.

[1] that would happen when prometheus-alert-buffer goes down & up and a client attempts to reconnect based on last known generation & index.

**Comment by simon3z** [comment]
@ilackarms @juliusv can you explicitly approve?

**Comment by juliusv** [comment]
:+1: Yes, looks good to me (I can't approve on GitHub, as I don't have any rights on this repo).

**Comment by simon3z** [comment]
Merging also according to a previous comment from @ilackarms
